### ISSUE TITLE
XamlIslands: Cannot create control of type

### ISSUE THREAD

**Issue by maoziliang** [microsoft/react-native-windows#10784]
### Problem Description

I'm investigating how to use XamlIslands to host the React Native Application. But it cannot startup successfully, the `WindowsXamlHost` element displays **Cannot create control of type xxx**
![image](https://user-images.githubusercontent.com/1581998/197323760-2ed2570e-9f90-4d34-b20a-6f51f3d54295.png)
![image](https://user-images.githubusercontent.com/1581998/197324788-c94cade7-24b0-4c70-9fd1-00598e2a0851.png)

The follow image shows my solution structure.
![image](https://user-images.githubusercontent.com/1581998/197323507-d0469b41-944e-4c7e-b89e-c53245c0756f.png)

1. Rn (Universal Windows). This project is used to provide an UWP control named `MyUserControl1` which uses `react:ReactRootView` to host the react native application.
2. RnApp (WPF .NET Core 3.1). This project will be deployed as my Application Entry Point. It will reference Rn (Universal Windows) and contains a page which uses  `MyUserControl1`.
3. RnAppPackage is which reference RnApp and will create a MSIX package.

### Steps To Reproduce

1. Init project. 
    1. npx react-native init rn --template react-native-template-typescript
    2. npx react-native-windows-init --overwrite --language cs --projectType lib --version 0.66
1. create WPF (.NET Core 3.1) project (named RnApp). And change the project file to add following lines.
    ```
    <AssetTargetFallback>uap10.0.19041</AssetTargetFallback>
    ```
1. Modify Class Library project(named rn) file. Add such lines to it.
    ```
    <PropertyGroup>
      <EnableTypeInfoReflection>false</EnableTypeInfoReflection>
      <EnableXBindDiagnostics>false</EnableXBindDiagnostics>
    </PropertyGroup>
    ```
1. create a **Windows Application Packaging Project** named RnAppPackage, set target version to 19041, and add **RnApp** as a reference.
1. In Rn project: Add new User Control named MyUserControl1 which will use ReactRootView to host the react native application.
1. In RnApp project: Open MainWindow.xaml, add `<xaml:WindowsXamlHost InitialTypeName="Rn.MyUserControl1" />`.
1. Build and start RnAppPackage

### Expected Results

The xaml:WindowsXamlHost display the UWP user control as expected.

### CLI version

6.4.0

### Environment

```markdown
info Fetching system and libraries information...
System:
    OS: Windows 10 10.0.19044
    CPU: (16) x64 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
    Memory: 43.92 GB / 63.20 GB
  Binaries:
    Node: 16.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 8.11.0 - C:\Program Files\nodejs\npm.CMD
    Watchman: Not Found
  SDKs:
    Android SDK: Not Found
    Windows SDK:
      AllowDevelopmentWithoutDevLicense: Enabled
      AllowAllTrustedApps: Enabled
      Versions: 10.0.18362.0, 10.0.19041.0
  IDEs:
    Android Studio: Not Found
    Visual Studio: 16.11.32630.194 (Visual Studio Community 2019)
  Languages:
    Java: Not Found
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.1.0 => 18.1.0
    react-native: ^0.66.0 => 0.66.4
    react-native-windows: 0.66.25 => 0.66.25
  npmGlobalPackages:
    *react-native*: Not Found
```

### Target Platform Version

10.0.19041

### Target Device(s)

Desktop

### Visual Studio Version

Visual Studio 2019

### Build Configuration

Debug

### Snack, code example, screenshot, or link to a repository

The following project can reproduce the issue.
https://github.com/maoziliang/rnInvestigation

**Comment by chrisglein** [comment]
Your project will need the reference to `XamlControlsResources` just like any WinUI project. I'm not sure in your configuration of WPF + Islands + toolkit where that goes. Do you have that somewhere in your project?

Also your version 0.66 is old and out of the support window. I don't think that's your problem here, but FYI you'll want to get up to date some point soon.

**Comment by jonthysell** [comment]
Reference will look something like this (taken from the non-islands RNW app template):

```xml
<Application.Resources>
  <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
</Application.Resources>
```

Not sure where to put it in your WPF + XamlIslands consumption scenario.

**Comment by maoziliang** [comment]
I add a new pure UWP ClassLibrary(named ClassLibrary1), it works well without add this `XamlControlsResources `block . 
I uploaded the code, very appreciate if you can check it.
![image](https://user-images.githubusercontent.com/1581998/197697122-31a13fc7-e86f-451b-b9d1-15aca63ad1fa.png)

**Comment by maoziliang** [comment]
I already uploaded the newest version to my [repo](https://github.com/maoziliang/rnInvestigation).

**Comment by maoziliang** [comment]
Hi, thanks very much for reminding me to add `XamlControlsResources `, I found it's my mistake. After carefully reading [this document](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/host-custom-control-with-xaml-islands). I managed to see the "Loading Bundle" screen of the React Application. But unfortunately, the application crashed after loading the whole bundle.

1. console logged
![image](https://user-images.githubusercontent.com/1581998/198209733-c72bc28a-4b85-4c8f-b8bd-90b4ebb5d889.png)

2. 
![image](https://user-images.githubusercontent.com/1581998/198209876-b296e9ad-e394-4b31-a1af-3775d700e1e9.png)

3. 
![image](https://user-images.githubusercontent.com/1581998/198209966-4fd4db2a-b318-4ddc-91b3-a7e30f113274.png)

**Comment by jonthysell** [comment]
It looks like your XamlIslands bug was resolved.

This new bug is different, and appears you are trying to use `useSyncExternalStore`, which is a React 18 API that is not available in React Native 0.66.

Please either remove your usage of that API or upgrade to React Native (Windows) 0.70 or greater (the first version built on top of React 18) and try again. Then if you still have a problem, please open a new bug.

**Comment by maoziliang** [comment]
thanks @jonthysell @chrisglein
